### DISCUSSION TITLE
Comparison with VSCode

### DISCUSSION THREAD

**Discussion by lcharles** [scalameta/coc-metals#373] [Q&A] [answered]
Hi, first I have to say thank you for this awesome vim plugin for scala.
I only have a couple of questions that I think are fair considering that its numbering suggests it will be reaching 1.0 milestone in a near future.

The first one is how do you compare it to the VSCode metals extension ? Maybe adding a Table in the Readme to describe the feature parity with popular editor VSCode and Intellij would be nice to offer users a quick way to choose. 
The second is about the completion results. When I use VSCode with metals, the completion of a function adds the parenthesis at the end of the function name and place the cursor inside them to allow users to type the parameters.
With coc-metals it only completes the name which is a bit frustrating to be honest. I was wondering if this was a limitation of coc itself, because I can report the same issue with the python extension.

Cheers,
L

**Comment by ckipp01** [comment, marked as answer]
Thanks for the kind words and question @lcharles.

> The first one is how do you compare it to the VSCode metals extension ? Maybe adding a Table in the Readme to describe the feature parity with popular editor VSCode and Intellij would be nice to offer users a quick way to choose.

This may not be exactly what you're looking for, but there is a [comparison chart](https://scalameta.org/metals/docs/editors/overview.html#editor-support) on the Metals site that does a decent job of showing the main features of Metals, and showing which editors support each. A full blown chart of every feature, especially with IntelliJ would probably be an extremely large task, and one I'm probably not going to put effort into. With that being said, the VS Code extension and coc-metals has a lot in common. They share  a common core ([metals-languageclient](https://github.com/scalameta/metals-languageclient)), and in general coc extensions are modeled after VS Code extensions. _Almost_ every Metals feature that is in the VS Code extension is in coc-metals, it's just they may look a bit different. A few examples:

- VS Code has great run/debug functionality built-in, whereas coc-metals relies on an external plugin for it, and at the moment it's limited to nvim.
- VS Code has inline implicit decorations. However, in Vim there is no such thing as virtual text, and in Nvim it's limited to the end of lines. So things like implicits are shown in the hover instead. 
- VS Code has a nice panel to show stack traces where you can just click on them and jump to the part of the code you need to go to. In coc-metals, we rely on it instead have the stack trace be copied from your bugger into a file Metals creates with code lenses that you need to trigger to go to that part of your code base.

In Metals we try quite hard to ensure that every feature we push out can be more or less implemented in any LSP editor. Although it does required the editor to implement some specific Metals LSP extensions.

>  The second is about the completion results. When I use VSCode with metals, the completion of a function adds the parenthesis at the end of the function name and place the cursor inside them to allow users to type the parameters.

So this is often overlooked, but this functionality exists in coc-metals as well. You just have to choose the correctly completion. I'll give you an example:

![2020-12-20 18 17 36](https://user-images.githubusercontent.com/13974112/102719689-b5399980-42ef-11eb-8a57-63caa0bdf5e6.gif)

Notice that they completions are quite similar. If you would select the first one, it would do what you're seeing locally probably, it would completion the method with the `()`, and your cursor behind the `()<cursor>`. Hover, you would need to select the second instead would would put the cursor where you'd want it, `(<cursor>)`.

**Reply by lcharles** [reply]
Hi ,thank you for you answer. I just realized that metals and coc-metals are from the same project so yeah it makes sense that this kind of information is there. My bad.
On the topic of completion you are right, it is working well with scala 2.13. But I think I still have this problem with Scala 3, is it expected ?

**Reply by ckipp01** [reply]
Completions for Scala 3 aren't nearly as polished as they are for Scala 3. Expect to see more improvements on them soon.

**Reply by lcharles** [reply]
Perfect, I created a few hello world programs to quickly test coc-metals and didn't realize that this time I was on Scala 3.
FYI, I was running it on Ubuntu in a vm but I plan on using it mainly on Windows. I don't think it really applies to this project but if it lacks testers for Windows I'd be happy to help.
